### ISSUE TITLE
Babun SSH Agent

### ISSUE THREAD

**Issue by nfarrar** [babun/babun#290]
I didn't see any information on configuring the ssh-agent with babun for key base authentication. Is there a recommended method?

**Comment by kubamarchwicki** [comment]
@airborn has shown it today with some non trivial amount of script code landing into .zshrc (or smth). Can't find the link but I'm sure Jacek will drop a reference here.

**Comment by airborn** [comment]
Basicly, I've used [this](https://help.github.com/articles/working-with-ssh-key-passphrases/#auto-launching-ssh-agent-on-msysgit) but added it yo my `.zshrc` instead of `.bashrc` 

The only inconvenience is that it ask for your password first time you open babun, not first time when you are trying to use your key.

**Comment by peppy** [comment]
The provided solution doesn't seem to work for me. It *does* prompt me for the passphrase on start-up, and doesn't on further window opens, but it also still asks me when connecting via ssh. Any further ideas?

**Comment by TheKas** [comment]
Put this in your .bash_profile

# ssh-agent Settings
SSHAGENT=/usr/bin/ssh-agent
SSH_AUTH_SOCK=$( ls -1 /tmp/ssh-*/* 2>/dev/null | head -n 1 )
if [ -z "$SSH_AUTH_SOCK" -a -x "$SSHAGENT" ]; then
        eval $($SSHAGENT -s)
        ssh-add ~/.ssh/id_rsa
        trap "kill $SSH_AGENT_PID" 0
else
    export SSH_AUTH_SOCK=${SSH_AUTH_SOCK}
    export SSH_AGENT_PID=$(ps -f|grep ssh-agent|awk '{print $2}')
fi

I know, this is quick an dirty, but it works!

**Comment by peppy** [comment]
This works wonders. Thanks!

**Comment by nfarrar** [comment]
I've been dorked around with this a little more. I shell sources this for cygwin environments. The function wraps ssh and auto-starts the pageant the first time you try to ssh to a host. It's not ideal, but I haven't been using windows enough to improve it yet.

    function ssh() {
      local _pageant_windows_path="C:\Program Files (x86)\PuTTY\pageant.exe"
      local _ppk_cygwin_path="$HOME/.ssh/id_rsa.ppk"
      local _pageant_cygwin_path="$(cygpath ${_pageant_windows_path})"
      local _ppk_windows_path="$(cygpath -w ${_ppk_cygwin_path})"
      local _ssh_pageant_path=$(which ssh-pageant)

      if [[ ! -x ${_pageant_cygwin_path} ]]; then
        echo "ERROR: Could not start ssh agent, pageant executable not found."
        return 1
      elif [[ ! -x ${_ssh_pageant_path} ]]; then
        echo "ERROR: Could not start ssh agent, ssh-pageant executable not found."
        return 1
      elif [[ ! -e ${_ppk_cygwin_path} ]]; then
        echo "ERROR: PPK file not present."
        return 1
      fi

      ps -W | grep -q 'pageant\.exe$'
      if [[ $? != 0 ]]; then
        echo "starting pageant ..."
        cygstart "${_pageant_cygwin_path}" "${_ppk_windows_path}"
      fi

      if [[ -z $SSH_AUTH_SOCK ]]; then
        eval $(/usr/bin/ssh-pageant -r -a "/tmp/.ssh-pageant-$(whoami)")
      fi

      command ssh $@
    }

**Comment by tombujok** [comment]
Could anyone describe in the FAQ and close this issue?

**Comment by juradoz** [comment]
I would add to @TheKas answer that it should be added to .zshrc if you`re using zsh, i.e. any vanilla babun install.

**Comment by reinoudk** [comment]
If you're using zsh you can also enable the ssh-agent plugin:
```
# ~/.zshrc
plugins=(git, ssh-agent)
```
https://github.com/robbyrussell/oh-my-zsh/blob/master/plugins/ssh-agent/ssh-agent.plugin.zsh

**Comment by kubamarchwicki** [comment]
+1
with for example pagent working on startup git works out of the box with babunshell

**Comment by felipemarques** [comment]
Thansk @TheKas !! Works perfectly!!!

**Comment by chukcha2** [comment]
@TheKas's solution works for me too. However I needed to make sure my aliased ls wasn't used since the aliased version (uses -F) appends "=" to socket files. To use unbiased version use "\ls".

**Comment by aairey** [comment]
Anyone who got this working with the KeePass plugin KeeAgent?

**Comment by skupfer** [comment]
@reinoudk works great. But if I limit the maximum lifetime and it expires it wont ask for the passphrase again. You have to manually kill the agent via task manager/console

**Comment by strarsis** [comment]
@aairey: Yes, as Babun is Cygwin based, the same steps/config applies to it as to Cygwin.

You let KeeAgent create a socket file to be used by Babun:
In KeePass (Menu ->)Tools -> KeeAgent -> Options...;
In window 'Options' -> Tab 'KeeAgent'; Section 'Cygwin/MSYS Integration':
[x] Create Cygwin compatible socket file (works with some versions of MSYS)
Path:/[Browse...] [Choose a destination file (e.g. 'D:\keeagent.sock')]
[OK]

Then you can access it over the special /cygdrive folder.

Add this to your .bashrc (assuming the KeeAgent socket file resides under D:\keeagent.sock):
````
# KeeAgent support
export SSH_AUTH_SOCK="/cygdrive/d/keeagent.sock"
````
Don't forget to source the .bashrc to apply it and it should work immediately.

Note that you have to unlock your workspace and, of course, ssh config has to be set up on Babun, too.

**Comment by aairey** [comment]
It is not working for me and the .sock files remains at 0 bytes in size ...
It seems that KeeAgent is not putting any data into it?

`ssh-add` then complains that the socket is not useable.

**Comment by strarsis** [comment]
@aairey: Is Agent Mode set to 'Auto'? Can KeeAgent write to socket file?
Try a different path. If still there is no socket file, it could be a bug.
On my system, the socket file is created when I close the 'Options' window with [OK].

**Comment by aairey** [comment]
@strarsis: Yes and yes. 
It does create the socket file, but also a shortcut file with `.lnk` extension.
Both are 0 bytes. maybe this is normal but babun/cygwin does not seem to like it.

**Comment by strarsis** [comment]
@aairey: This seems to be a bug in or an issue with KeeAgent, you should create an issue: https://github.com/dlech/KeeAgent/issues

**Comment by nmeyer-otto** [comment]
To add SSH Keys on demand, upgrade ssh to at least version 7.2 and use the ssh-agent plugin for zsh and add the following settings:
```
pact update openssh
zstyle :omz:plugins:ssh-agent identities '' # doesn't add any keys on startup
```

Add the following line to your ssh config (.ssh/config, make sure permissions are 600)
```
AddKeysToAgent yes
```

**Comment by aairey** [comment]
I closed my issue with KeeAgent. it is working perfectly.

The simplest method would indeed be to use the `ssh-agent` oh-my-zsh plugin.
This does everything for you.

If you use KeeAgent as ssh-agent, you need to set the `SSH_AUTH_SOCK` environment variable _after_ all oh-my-zsh scripts have loaded (so at the very end of your `.zshrc`).

**Comment by slefevre** [comment]
Babun seems to use zsh out of the box, so I was confounded when I added these lines to my .bash_profile file. Instead, of course, I needed to add the lines to my .zsh_profile file.
